Modules/gcmodule.c

```diff
@@ -378,15 +444,14 @@ move_unreachable(PyGC_Head *young, PyGC_Head *unreachable)
              */
             PyObject *op = FROM_GC(gc);
             traverseproc traverse = Py_TYPE(op)->tp_traverse;
-            assert(_PyGCHead_REFS(gc) > 0);
-            _PyGCHead_SET_REFS(gc, GC_REACHABLE);
+            assert(gc_get_refs(gc) > 0);
             (void) traverse(op,
-                            (visitproc)visit_reachable,
-                            (void *)young);
+                    (visitproc)visit_reachable,
+                    (void *)young);
+            gc_set_prev(gc, prev);
+            gc->gc.gc_prev &= ~MASK_COLLECTING;
+            prev = gc;
             next = gc->gc.gc_next;
-            if (PyTuple_CheckExact(op)) {
-                _PyTuple_MaybeUntrack(op);
-            }
```

Comment:
Since `gc->gc.gc_next->gc.gc_prev` is not real prev pointer, we can't call `_PyTuple_MaybeUntrack(op)` here.
I moved it to split function.  But it means this pull request adds one more link traversal.

By the way, current code is not idiomatic.  While most tuples can be untracked, there are still many tuples
which arn't be untracked.  For example, `__mro__` or `__bases__`.
In final generation, we check all tuples contents.  In applications which triggers final generation GC frequently,
and there are many tuples, it can be significant overhead.

Anyway, I need to find "tracked tuple heavy" application to benchmark before optimize here.

Reply:
If final generation GC is triggered frequently, it means our heuristic is inadequate. Full collections can be very expensive.